theano/gpuarray/subtensor.py

```diff
@@ -447,6 +447,121 @@ def c_code_cache_version(self):
+            if isinstance(i, numpy.ndarray):
+                start = k
+                break
+        for k, i in enumerate(idx[::-1]):
+            if isinstance(i, numpy.ndarray):
+                end = len(idx) - k
+                break
+        # step 2: transpose
+        def get_indices(a, b, ind):
+            """
+            Get real indices for a list of indices.
+            """
+            dimshuffle_info = []
+            new_ind = []
+            k = 0
+
+            for i in range(0, a):
+                if isinstance(ind[i], slice):
+                    dimshuffle_info.append(k)
+                    new_ind.append(ind[i])
+                    k += 1
+                elif ind[i] is None:
+                    dimshuffle_info.append('x')
+                    new_ind.append(slice(None))
+
+            dimshuffle_info.append(k)
+            new_ind.append(ind[a])
+            k += 1
+
+            idx_1 = []
+            idx_2 = []
+            idx_3 = []
+            for i in range(a + 1, b):
+                if isinstance(ind[i], slice):
+                    idx_1.append(k)
+                    idx_2.append(ind[i])
+                    k += 1
+                elif ind[i] is None:
+                    idx_3.append('x')
+                    new_ind.append(slice(None))
+                else:
+                    idx_3.append(k)
+                    new_ind.append(ind[i])
+                    k += 1
+            valid_end = a + len(idx_3) + 1
+
+            dimshuffle_info.extend(idx_3)
+            new_ind += idx_2
+            dimshuffle_info.extend(idx_1)
+
+            for i in range(b, len(ind)):
+                if isinstance(ind[i], slice):
+                    dimshuffle_info.append(k)
+                    new_ind.append(ind[i])
+                    k += 1
+                elif ind[i] is None:
+                    dimshuffle_info.append('x')
+                    new_ind.append(slice(None))
+
+            return dimshuffle_info, new_ind, valid_end
+
+        (dimshuffle_idx, new_ind,
+         end_) = get_indices(start, end, idx)
+        x = x.transpose(*dimshuffle_idx)
+        # step 3: partial flattening
+        start_ = start
+        shape = (x.shape[: start_] +
+                 (numpy.prod(x.shape[start: end_]),) +
+                 x.shape[end_:])
+        input_flat = numpy.reshape(x, shape)
+        # step 4: build the strides
+        strides = [1]
+        for i in range(start_, end_ - 1)[::-1]:
+            stride = x.shape[i + 1] * strides[-1]
+            strides.append(stride)
+        # step 5: build the indices into x_flat
+        items = [new_ind[i] if isinstance(new_ind[i], numpy.ndarray)
+                 else 0 for i in range(start_, end_)]
+        new_idx = numpy.sum([i * j for i, j
+                             in zip(items, strides[::-1])],
+                            axis=0)
+        # step 6: advanced slicing
+        out_flat = input_flat.take(new_idx.flatten())
```

Comment:
I just discovered that GpuArrays don't expose `take()`.  So we should probably do that.  I'll have a PR for that tomorrow.